### PR TITLE
[12.x] Optimize + simplify blade foreach loops

### PR DESCRIPTION
**Pull request by bert-w**
The compilation of `@foreach` blade directives in `<=11.x` results in pretty verbose code to make the `$loop` variable work. This PR greatly simplifies that output while improving the general logic of the `$loop` variable.

This changes the output of a compiled blade file from:
```php
<ul>
<?php $__currentLoopData = [1, 2]; $__env->addLoop($__currentLoopData); foreach($__currentLoopData as $outer): $__env->incrementLoopIndices(); $loop = $__env->getLastLoop(); ?>
    <li>i: <?php echo e($outer); ?>
        <ul>
            <?php $__currentLoopData = [1, 2, 3]; $__env->addLoop($__currentLoopData); foreach($__currentLoopData as $inner): $__env->incrementLoopIndices(); $loop = $__env->getLastLoop(); ?>
                <li>j: <?php echo e($inner); ?></li>
            <?php endforeach; $__env->popLoop(); $loop = $__env->getLastLoop(); ?>
        </ul>
    </li>
<?php endforeach; $__env->popLoop(); $loop = $__env->getLastLoop(); ?>
</ul>
<?php /**PATH C:\...\loops.blade.php ENDPATH**/ ?>
```

To:
```php
<ul>
<?php foreach($__env->addLoop([1, 2]) as $outer): $loop = $__env->getLastLoop(); ?>
    <li>i: <?php echo e($outer); ?>
        <ul>
            <?php foreach($__env->addLoop([1, 2, 3]) as $inner): $loop = $__env->getLastLoop(); ?>
                <li>j: <?php echo e($inner); ?></li>
            <?php $__env->incrementLoopIndices(); endforeach; $loop = $__env->popLoop(); ?>
        </ul>
    </li>
<?php $__env->incrementLoopIndices(); endforeach; $loop = $__env->popLoop(); ?>
</ul>
<?php /**PATH C:\...\loops.blade.php ENDPATH**/ ?>
```

Noteworthy changes:
- The `$__currentLoopData` variable no longer exists
- `$__env->addLoop()` now returns the input data, so it can be used instantly within the `foreach` struct
- `$__env->popLoop()` now returns the new loop variable immediately
- The `$__env->incrementLoopIndices()` is now moved to the end of each iteration, since the initial `$loop` variable assignment (the `$loop = $__env->getLastLoop();`) will now contain correct data from the start (the previous implementation created an incomplete placeholder which had to be incremented in order to be correct). The new implementation should have no behavioral change whatsoever.
- The changes in the tests correspond to the fact that the initial loop variable is now correct.

Here is a simple blade test script to check the output of the loop variables (even though it is also covered by the tests):
- https://gist.github.com/bert-w/b818cddfba8f02a2f271181373217120

I have submitted this to the master branch because **this change requires a recompilation of all views** (so a `php artisan view:clear` would be required).

### CODE REVIEW

File: tests/View/ViewFactoryTest.php [hunk 1]
```diff
@@ -917,14 +917,14 @@ public function testAddingLoops()
         $factory->addLoop([1, 2, 3]);
 
         $expectedLoop = [
-            'iteration' => 0,
+            'iteration' => 1,
```
Comment (jmarcher) [inline tests/View/ViewFactoryTest.php:920]: This should not change, why the change?

Maybe this is a big breaking change after all

**Comment by jmarcher** [comment]
I ask myself what is the benefit from this changes? Compiled blades are not meant to be pretty, or does this change brings a big performance improvement?

File: tests/View/ViewFactoryTest.php (hunk 1 above)
Comment (bert-w) [reply tests/View/ViewFactoryTest.php:920]: This should function exactly the same. What previously happened is that the code did some initialization with `iteration=0`, after which it immediately incremented the step directly after the `foreach()` struct, thus ending up at `iteration=1` anyway. This extra step is not necessary because you can initialize it correctly immediately.

To address your other comment:

- This PR simplifies the logic thus making the compiled files more readable (compiled view files are often referenced in exceptions so you will end up in these files one way or another)
- It saves a few CPU cycles
- It is mostly old code from 2016 which could use an update

**Comment by KennedyTedesco** [comment]
Your changes look good overall, but I agree with @jmarcher, this only makes sense if it brings a significant performance improvement. The compiled view is something people don't touch, so it doesn't need to be less verbose.

**Comment by taylorotwell** [comment]
Hmm, yeah - think I'll skip this one unless it's a really big perf boost.
